File: src/main/java/juuxel/woodsandmires/feature/MirePondsFeature.java
package juuxel.woodsandmires.feature;

import com.mojang.serialization.Codec;
import net.minecraft.block.BlockState;
import net.minecraft.block.Blocks;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.math.Direction;
import net.minecraft.world.Heightmap;
import net.minecraft.world.StructureWorldAccess;
import net.minecraft.world.gen.feature.DefaultFeatureConfig;
import net.minecraft.world.gen.feature.Feature;
import net.minecraft.world.gen.feature.util.FeatureContext;

public class MirePondsFeature extends Feature<DefaultFeatureConfig> {
    private static final Direction[] NON_UP_DIRECTIONS = new Direction[] {
        Direction.NORTH,
        Direction.EAST,
        Direction.SOUTH,
        Direction.WEST,
        Direction.DOWN
    };

    public MirePondsFeature(Codec<DefaultFeatureConfig> configCodec) {
        super(configCodec);
    }

    @Override
    public boolean generate(FeatureContext<DefaultFeatureConfig> context) {
        var world = context.getWorld();
        var pos = context.getOrigin();
        var config = context.getConfig();
        var random = context.getRandom();

        BlockState water = Blocks.WATER.getDefaultState();
        BlockPos.Mutable mut = new BlockPos.Mutable();
        boolean generated = false;

        for (int x = 0; x < 16; x++) {
            for (int z = 0; z < 16; z++) {
                // Leave holes in the ponds
                if (random.nextBoolean()) continue;

                int xo = pos.getX() + x;
                int zo = pos.getZ() + z;
                int y = world.getTopY(Heightmap.Type.MOTION_BLOCKING, xo, zo) - 1;
                mut.set(xo, y, zo);

                if (isSoil(world, mut) && isSolidOrWaterAround(world, mut)) {
                    setBlockState(world, mut, water);
                    generated = true;

                    if (random.nextInt(3) == 0) {
                        mut.move(Direction.DOWN);

                        if (world.getBlockState(mut).isSolidBlock(world, mut) & isSolidOrWaterAround(world, mut)) {
                            setBlockState(world, mut, water);
                        }
                    }
                }
            }
        }

        return generated;
    }

    private static boolean isSolidOrWaterAround(StructureWorldAccess world, BlockPos.Mutable pos) {
        for (Direction direction : NON_UP_DIRECTIONS) {
            pos.move(direction);
            BlockState state = world.getBlockState(pos);
            pos.move(direction.getOpposite());

            if (!state.isOf(Blocks.WATER) && !state.isSolidBlock(world, pos)) {
                return false;
            }
        }

        return true;
    }
}

File: src/main/kotlin/juuxel/woodsandmires/feature/MirePondsFeature.kt
package juuxel.woodsandmires.feature

import com.mojang.serialization.Codec
import net.minecraft.block.Blocks
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.world.Heightmap
import net.minecraft.world.ServerWorldAccess
import net.minecraft.world.gen.StructureAccessor
import net.minecraft.world.gen.chunk.ChunkGenerator
import net.minecraft.world.gen.feature.DefaultFeatureConfig
import net.minecraft.world.gen.feature.Feature
import java.util.Random

class MirePondsFeature(configCodec: Codec<DefaultFeatureConfig>) : Feature<DefaultFeatureConfig>(configCodec) {
    override fun generate(
        world: ServerWorldAccess, structureAccessor: StructureAccessor, generator: ChunkGenerator,
        random: Random, pos: BlockPos, config: DefaultFeatureConfig
    ): Boolean {
        val water = Blocks.WATER.defaultState
        val mut = BlockPos.Mutable()
        var generated = false

        for (x in 0..15) {
            for (z in 0..15) {
                if (random.nextBoolean()) continue

                val xo = pos.x + x
                val zo = pos.z + z
                mut.set(xo, world.getTopY(Heightmap.Type.MOTION_BLOCKING, xo, zo) - 1, zo)

                if (method_27368(world, mut) && isSolidOrWaterAround(world, mut)) {
                    world.setBlockState(mut, water, 2)
                    generated = true

                    if (random.nextInt(3) == 0) {
                        mut.move(0, -1, 0)
                        if (method_27368(world, mut) && isSolidOrWaterAround(world, mut)) {
                            world.setBlockState(mut, water, 2)
                        }
                    }
                }
            }
        }

        return generated
    }

    private fun isSolidOrWaterAround(world: ServerWorldAccess, pos: BlockPos.Mutable): Boolean =
        Direction.values().all { direction ->
            if (direction == Direction.UP) return@all true

            pos.move(direction)
            val state = world.getBlockState(pos)
            pos.move(direction.opposite)

            state.isOf(Blocks.WATER) || state.isSolidBlock(world, pos)
        }
}
